## 6.3 説明変数が2個以上の場合の解析方法

説明変数として、サークルに所属しているか否かという情報を加える。  
これまでと同様に、所属しているか否かについて、ダミー変数で  
表現してやれば良い。  
やはり、線形代数の成績カテゴリーの1つである「優」の項目を  
削除したモデルを使って解析を行う。ちなみに、「優」の項目を  
削除した成績表はcsv.6.3のようになる。

## 以降はstatsmodelを使う

In [2]:
#ライブラリをインポート
import pandas as pd
import numpy as np
%matplotlib inline
import matplotlib.pyplot as plt
import math
import statsmodels.api as sm

In [3]:
#データフレームの作成
df = pd.read_csv("6.3.csv", sep=",")

In [5]:
#説明変数を取り出す
X = df.loc[:, ['x12','x13','x2']].values

#ターゲット変数を取り出す
y = df['score'].values

In [6]:
#モデルの作成、バイアス項を追加
model = sm.OLS(y, sm.add_constant(X))
results = model.fit()

#結果を表示
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.818
Model:                            OLS   Adj. R-squared:                  0.727
Method:                 Least Squares   F-statistic:                     9.007
Date:                Mon, 24 Aug 2020   Prob (F-statistic):             0.0122
Time:                        20:37:40   Log-Likelihood:                -28.236
No. Observations:                  10   AIC:                             64.47
Df Residuals:                       6   BIC:                             65.68
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         83.0000      3.143     26.405      0.0

/opt/anaconda3/lib/python3.7/site-packages/scipy/stats/stats.py:1450: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=10
  "anyway, n=%i" % int(n))


In [7]:
#重相関係数を計算
R = math.sqrt(0.818)
print(R)

0.904433524367601


Rを計算するために、表の値を書き写してしまっているが、上の表からほしい統計量を取り出す方法が  
あれば教えてほしい。

In [11]:
#予測値の書き出し
pred = results.predict(sm.add_constant(X))
print(pred)

[92. 92. 83. 83. 82. 73. 73. 73. 73. 64.]


## 変数選択

説明変数として線形代数の試験結果を選択した場合のF_0値は5.51であった。  
次に、説明変数としてサークルに所属しているか否かの情報を選択した場合の  
F_0値を調べる。Club activityのスクリプトを参照。  
サークルの情報もモデルに組み込まなかった場合と組み込んだ場合について  
adjusted R^2を比較すると、それぞれ0.500, 0.727であり、後者の方が  
改善されていることがわかる。従い、サークルの情報はモデルの含めるべき  
であることがわかる。